### PR TITLE
Add AWS Network Firewall logging, tagging functionality

### PR DESCRIPTION
**Pull request by jerryhxu**
This PR accomplishes the following:

1. Closes [issue 9327](https://github.com/cloud-custodian/cloud-custodian/issues/9327)
2. Adds tagging functionality to bedrock-agent resource.
3. Fixes 1 issue with s3 bucket: when permission is denied to get replication rule, code can return nothing.
4. Adds annotation-key for model invocation logging so that the logging config data can be returned in json once the filter is applied.

Note: 'resourcegroupstaggingapi' cannot tag network firewall and bedrock-agent resource. It returns error:
   `Unrecognized service or resource type for tagging`
That is why additional tagging functionality is implemented.

### CODE REVIEW

**Review by ajkerrigan — COMMENTED** [review]
Thanks @jerryhxu! This looks good to me for the most part :+1: . Some comments/suggestions inline, and I think going forward changes like this could make sense in separate PRs. It would be easier to get the smaller bits/fixes merged while we work through questions about the new actions.

File: c7n/resources/firewall.py [hunk 1]
```diff
@@ -61,3 +67,273 @@ class FirewallVpcFilter(VpcFilter):
...
+                                'LogDestinationConfigs': [loggingConfigurations[0]]
+                                }
+                            )
+                        client.update_logging_configuration(
+                            FirewallName=r['FirewallName'],
+                            FirewallArn =r['FirewallArn'],
+                            LoggingConfiguration = { 'LogDestinationConfigs': []}
+                            )
+                    except client.exceptions.ResourceNotFoundException:
+                        continue
```
[the remark this answers is not in the document]
Comment (jerryhxu) [reply c7n/resources/firewall.py]: yeah, I agree this is much cleaner. Thanks for the suggestion!

File: c7n/resources/firewall.py [hunk 2]
```diff
@@ -61,3 +67,273 @@ class FirewallVpcFilter(VpcFilter):
...
+
+        policies:
+          - name: network-firewall-delete
+            resource: aws.firewall
+            actions:
+              - type: delete
+    """
+    schema = type_schema('delete',
+        force={'type': 'boolean', 'default': False})
+    permissions = ('network-firewall:DeleteFirewall',)
```
[the remark this answers is not in the document]
Comment (jerryhxu) [reply c7n/resources/firewall.py]: I missed this part. Thanks for correcting it.

File: c7n/resources/firewall.py [hunk 3]
```diff
@@ -61,3 +67,273 @@ class FirewallVpcFilter(VpcFilter):
...
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('network-firewall')
+        if self.data.get('force', False):
+            del_protection_updater = self.manager.action_registry['update-delete-protection'](
+                {'type': 'update-delete-protection', 'state': False}, self.manager)
+            del_protection_updater.process(resources)
+            logging_updater = self.manager.action_registry['update-logging-config'](
+                {'type': 'update-logging-config', 'enabled': False}, self.manager)
```
[the remark this answers is not in the document]
Comment (jerryhxu) [reply c7n/resources/firewall.py:209]: It seems that existing codebase uses the level of indirection such as: [ ](https://github.com/cloud-custodian/cloud-custodian/blob/main/c7n/resources/dynamodb.py#L231-L234). There are a number of other examples as well. I will keep it this way to be consistent.

File: c7n/resources/firewall.py [hunk 4]
```diff
@@ -61,3 +67,273 @@ class FirewallVpcFilter(VpcFilter):
...
+              self.service)
+
+    # Base on AWS documentation:
+    # You can perform only one of the following actions in any call to UpdateLoggingConfiguration:
+    # Create a new log destination object by adding a single LogDestinationConfig array element
+    #    to LogDestinationConfigs.
+    # Delete a log destination object by removing a single LogDestinationConfig array element from
+    #     LogDestinationConfigs.
+    # Change the LogDestination setting in a single LogDestinationConfig array element.
+    def process(self, resources):
```
[the remark this answers is not in the document]
Comment (jerryhxu) [reply c7n/resources/firewall.py:307]: Not sure why AWS did it this way. I hope they change it in the future.

**Review by ajkerrigan — APPROVED** [review]
Thanks for the updates, looks good! :+1:

**Pull request merged**